langserver/lint.go

```diff
@@ -0,0 +1,199 @@
+// results as diagnostics.
+func (h *LangHandler) lint(ctx context.Context, bctx *build.Context, conn jsonrpc2.JSONRPC2, args []string, files []string) error {
+	if h.linter == nil {
+		return nil
+	}
+
+	diags, err := h.linter.Lint(ctx, bctx, args...)
+	if err != nil {
+		return err
+	}
+
+	return h.publishDiagnostics(ctx, conn, diags, h.config.LintTool, files)
+}
+
+// lintPackage runs LangHandler.lint for the package containing the uri.
+func (h *LangHandler) lintPackage(ctx context.Context, bctx *build.Context, conn jsonrpc2.JSONRPC2, uri lsp.DocumentURI) error {
+	pkg, err := ContainingPackage(h.BuildContext(ctx), h.FilePath(uri))
+	if err != nil {
+		return err
+	}
+
+	files := make([]string, 0, len(pkg.GoFiles))
+	for _, f := range pkg.GoFiles {
+		files = append(files, path.Join(pkg.Dir, f))
+	}
+
+	return h.lint(ctx, bctx, conn, []string{pkg.ImportPath}, files)
+}
+
+// lintWorkspace runs LangHandler.lint for the entire workspace
+func (h *LangHandler) lintWorkspace(ctx context.Context, bctx *build.Context, conn jsonrpc2.JSONRPC2) error {
+	rootPkg, err := ContainingPackage(h.BuildContext(ctx), h.RootFSPath)
+	if err != nil {
+		return err
+	}
+
+	var files []string
+	pkgs := tools.ListPkgsUnderDir(bctx, h.RootFSPath)
+	for _, pkg := range pkgs {
+		p, err := bctx.Import(pkg, h.RootFSPath, 0)
+		if err != nil {
+			return err
+		}
+
+		for _, f := range p.GoFiles {
+			files = append(files, path.Join(p.Dir, f))
+		}
+	}
+
+	return h.lint(ctx, bctx, conn, []string{path.Join(rootPkg.ImportPath, "/...")}, files)
+}
+
+// golint is a wrapper around the golint command that implements the
+// linter interface.
+type golint struct{}
+
+func (l golint) IsInstalled(ctx context.Context, bctx *build.Context) error {
+	cmd := exec.CommandContext(ctx, "golint", "--help")
+	cmd.Env = []string{
+		"GOPATH=" + bctx.GOPATH,
+		"GOROOT=" + bctx.GOROOT,
+	}
+
+	buff := new(bytes.Buffer)
+	cmd.Stdout = buff
+	cmd.Stderr = buff
+
+	err := cmd.Run()
+
+	if err == nil {
+		return nil
+	}
+	if _, ok := err.(*exec.ExitError); ok {
+		return nil
+	}
+
+	return fmt.Errorf("linter check failed: %s (%q)", err, buff.String())
+}
+
+func (l golint) Lint(ctx context.Context, bctx *build.Context, args ...string) (diagnostics, error) {
+	cmd := exec.CommandContext(ctx, "golint", args...)
+	cmd.Env = []string{
+		"GOPATH=" + bctx.GOPATH,
+		"GOROOT=" + bctx.GOROOT,
+	}
+
+	buff := new(bytes.Buffer)
+	cmd.Stdout = buff
+	cmd.Stderr = buff
+
+	err := cmd.Run()
+	if err != nil {
+		switch err.(type) {
+		case *exec.ExitError:
+			// do nothing
+		default:
+			return nil, fmt.Errorf("lint command error: %s", err)
+		}
+	}
+
+	diags := diagnostics{}
+	for {
+		l, err := buff.ReadString('\n')
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("could not read lint command output: %s", err)
+		}
+
+		l = strings.TrimSuffix(l, "\n")
+		file, line, _, message, err := parseLintResult(l)
```

Comment:
you include stderr in buff which probably doesn't conform to the expected input to parseLintResult. parseLintResult also isn't very forgiving. Maybe you shouldn't include stderr and just log it separately?

Additionally there isn't really a reason to read this all into a bytes buffer. Scanner is good at reading an io.Reader line by line.